### PR TITLE
fix: large old replay data splitting

### PR DESCRIPTION
**Pull request by pauldambra**
we see a non-zero amount of MessageTooLarge errors

these are all (so far on sampling inspection of the data) from old clients that didn't have batching code and would sometimes send a lot of data in one go

around 1 in 5 of them have many hundreds of items to process 
1 in 25 has tens of thousands of items

we already have code that should be splitting these out into individual events

but clearly it's not working

and we don't really want one API call to generate 10k kafka messages 🙈 

so this PR

### changes how we check the headroom - we're clearly under counting, this might help

i looked at how the data is going to be sent to kafka and tried to copy that so that we're counting bytes, and counting a similar bytes array, instead of counting characters, JS in the browser uses UTF-16 string and kafka/python is using UTF-8 so maybe there's some silliness happening here

### split the list instead of exploding it

the final case in the processing if the non-full snapshots won't fit into headroom sends every item from the list individually

instead now, we keep splitting the list into 2 and checking the size of each half
in theory this means the majority case is we'll split into one or two messages each with many events

### CODE REVIEW

File: posthog/session_recordings/session_recording_helpers.py [hunk 1]
```diff
@@ -202,10 +206,25 @@ def new_event(items: list[dict] | None = None) -> Event:
...
+                    new_parts = []
+                    for part in parts:
+                        if byte_size_dict(part) < size_with_headroom:
+                            event = new_event(part)
+                            yield event
+                        elif len(part) > 1:  # Only split if the part has more than one item
+                            part1, part2 = split_list(part)
+                            new_parts.extend([part1, part2])
+                        else:
+                            # If part has only one item and still doesn't fit, yield it as is
```
Comment (pauldambra) [inline posthog/session_recordings/session_recording_helpers.py:224]: i believe this stops us looping indefinitely

Comment (pauldambra) [reply posthog/session_recordings/session_recording_helpers.py:223]: actually, i'm going to put an artificial limit on this loop for safety

File: posthog/session_recordings/session_recording_helpers.py [hunk 2]
```diff
@@ -202,10 +206,35 @@ def new_event(items: list[dict] | None = None) -> Event:
...
-                # If not, send them individually
-                for snapshot_data in other_snapshots:
-                    event = new_event([snapshot_data])
-                    yield event
+                # If not, keep splitting the list until all parts fit within headroom
+                # we want to avoid sending them all individually if we can - there could be tens of thousands
+                # in data from these older clients that batched poorly
+                parts = [other_snapshots]
+                loop_count = 0
+                while parts and loop_count < 10:
```
Comment (daibhin) [inline posthog/session_recordings/session_recording_helpers.py:214]: Maybe make this a constant that makes it clear it is an arbitrary number that can be played with

**Review by daibhin — APPROVED** [review]
Nice!!

**Comment by pauldambra** [comment]
removes the test here to get the PR moving instead of waiting for https://github.com/PostHog/posthog/pull/23466

**Pull request merged**